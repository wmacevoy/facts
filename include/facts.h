#pragma once


#ifdef __cplusplus
#include <iostream>
#endif

#include <stdio.h>
#include <string.h>


struct FactsStruct;
typedef struct FactsStruct Facts;

#define FACTS_SIG {0xe3,0xb0,0xc4,0x42, 0x98,0xfc,0x1c,0x14, 0x9a,0xfb,0xf4,0x89, 0x96,0xfb,0x92,0x00}

struct FactsStruct {
  unsigned char sig[16];
  const char *file;
  int line;
  const char *name;
  void (*function)(Facts *facts);
  int status;
  Facts *next;
  Facts *prev;
};

void FactsPrint(const char *fmt,...);
void FactsFindInMemory(unsigned char *begin,unsigned char *end);
void FactsInclude(const char *pattern);
void FactsExclude(const char *pattern);
void FactsCheck();
void FactsFiction(const char *file, int line,Facts *facts);

double FactsAbsErr(double a, double b);
double FactsRelErr(double a, double b);

int FactsMain(int argc, const char *argv[]);

// https://stackoverflow.com/questions/24844970/how-to-print-types-of-unknown-size-like-ino-t
#define FACTS_PRINT_FORMAT(X) _Generic((X),				\
				       char: "%c",			\
				       unsigned char: "%hhu",		\
				       unsigned short: "%hu",		\
				       unsigned int: "%u",		\
				       unsigned long: "%lu",		\
				       unsigned long long: "%llu",	\
				       signed char: "%hhd",		\
				       short: "%hd",			\
				       int: "%d",			\
				       long: "%ld",			\
				       long long: "%lld",		\
				       float: "%g",			\
				       double: "%g",			\
				       long double: "%Lg",		\
				       const char *: "%s",		\
				       const unsigned char *: "%s",	\
				       const void *: "%p",		\
				       char *: "%s",			\
				       unsigned char *: "%s",		\
				       void *: "%p")
  
#ifndef FACTS_C
extern int facts_fictions;
extern int facts_truths;
#endif

#define CHECK_PRINT(a,op,b,fmt) (((a) op (b)) ? (++facts_truths,1) : (FactsFiction(__FILE__,__LINE__,facts),FactsPrint("%s %d: %s {=%?} " #op " %s {=%?} is fiction\n",fmt,fmt,__FILE__,__LINE__,#a,(a),#b,(b)), facts->status=-1,0))
#define FACT_PRINT(a,op,b,fmt) if (!CHECK_PRINT(a,op,b,fmt)) return

#define CHECK_CERR(a,op,b) (((a) op (b)) ? (++facts_truths,1) : (FactsFiction(__FILE__,__LINE__,facts),std::cerr << __FILE__ << " " << __LINE__ << ": " << #a << "{=" << (a) << "} " << #op << #b << " {=" << (b) << "} is fiction" << std::endl, status=-1,0))
#define FACT_CERR(a,op,b) if (!CHECK_CERR(a,op,b)) return


#ifdef __cplusplus
#define FACT(a,op,b) FACT_CERR(a,op,b)
#define CHECK(a,op,b) CHECK_CERR(a,op,b)
#else
#define CHECK(a,op,b) CHECK_PRINT(a,op,b,FACTS_PRINT_FORMAT(a))
#define FACT(a,op,b) FACT_PRINT(a,op,b,FACTS_PRINT_FORMAT(a))
#endif

#define FACTS(name) void facts_ ##name## _function (Facts *facts); Facts facts_ ##name## _data = { FACTS_SIG, __FILE__, __LINE__, #name, &facts_ ##name## _function, 0, NULL, NULL }; void facts_ ##name## _function(Facts *facts)

#ifndef FACTS_C
FACTS(0000_BEGIN) {}
#endif

#define FACTS_FINISHED FACTS(zzzz_END) {}; void FactsRegister() { FactsFindInMemory((unsigned char*)&facts_0000_BEGIN_data, (unsigned char *)&facts_zzzz_END_data); }

#define FACTS_MAIN int main(int argc, const char *argv[]) { return FactsMain(argc, argv); }
