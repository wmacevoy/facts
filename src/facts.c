#include <stdarg.h>
#include <math.h>
#include <assert.h>
#include <stdint.h>

#define FACTS_C 1
#include "facts.h"
static Facts *head = NULL, *tail=NULL;

int facts_fictions = 0;
int facts_truths = 0;
void FactsRegister();

static int matches(const char *str, const char *pattern) {
  uint8_t np = strlen(pattern);
  uint8_t ns = strlen(str);
  uint8_t nb = (np+2)/8+((np+2)%8 != 0);
  uint8_t k;
  uint8_t buffer[8];

  uint8_t buffer0[nb];
  uint8_t buffer1[nb];

  uint8_t *state0=buffer0;
  uint8_t *state1=buffer1;

  memset(state0,0,nb);
  state0[0]=1;
  for (k=1; pattern[k-1] == '*'; ++k) state0[k/8] |= (1 << (k%8));

  for (int i=0; i<=ns; ++i) {
    uint8_t c=str[i];

    memset(state1,0,nb);
    for (int j=0; j<=np; ++j) {
      if (state0[j/8] & (1 << (j%8))) {
        if (pattern[j] == '*') {
          k=j;
          state1[k/8] |= (1 << (k%8));
          ++k;
          state1[k/8] |= (1 << (k%8));
        } else if (pattern[j] == c) {
          k=j+1;
          state1[k/8] |= (1 << (k%8));
          while (pattern[k] == '*') {
            ++k; 
            state1[k/8] |= (1 << (k%8));
          }
        }
      }
    }

    uint8_t *tmp = state0;
    state0 = state1;
    state1 = tmp;
  }

  k=np+1;
  return (state0[k/8]&(1<<(k%8))) != 0;
}

void FactsFiction(const char *file, int line, Facts *facts) {
  ++facts_fictions;
}


void FactsInclude(const char *pattern) {
  if (head == NULL) { FactsRegister(); }  
  for (Facts *facts = head; facts != NULL; facts=facts->next) {
    if (matches(facts->name,pattern)) {
      facts->status = 0;
    }
  }
}

void FactsExclude(const char *pattern) {
  if (head == NULL) { FactsRegister(); }  
  for (Facts *facts = head; facts != NULL; facts=facts->next) {
    if (matches(facts->name,pattern)) {
      facts->status = -2;
    }
  }
}

void FactsFindInMemory(unsigned char *begin, unsigned char *end) {
  int reversed = 0;
  if (end < begin) {
    void *tmp = end;
    end = begin;
    begin = tmp;
    reversed = 1;
  }
  end = end - 16;
  
  unsigned char *p = (unsigned char*) begin;
  for (unsigned char *p = begin; p != NULL && p < end; p = (unsigned char*) memchr(p+16,0xe3,end-p)) {
    if (p[ 0]==0xe3 && p[ 1]==0xb0 && p[ 2]==0xc4 && p[ 3]==0x42 &&
	p[ 4]==0x98 && p[ 5]==0xfc && p[ 6]==0x1c && p[ 7]==0x14 &&
	p[ 8]==0x9a && p[ 9]==0xfb && p[10]==0xf4 && p[11]==0x89 &&
	p[12]==0x96 && p[13]==0xfb && p[14]==0x92 && p[15]==0x00) {

      Facts *facts = (Facts*)p;

      if (facts->name != NULL && facts->function != NULL && facts->prev == NULL && facts->next == NULL) {
	if (strcmp(facts->name,"0000_BEGIN")==0) continue;
	if (strcmp(facts->name,"zzzz_END")==0) continue;
	if (reversed) {
	  facts->next = head;
	  facts->prev = NULL;
	  if (head != NULL) { head->prev = facts; }
	  if (tail == NULL) { tail = facts; }
	  head = facts;
	} else {
	  facts->prev = tail;
	  facts->next = NULL;
	  if (tail != NULL) { tail->next = facts; }
	  if (head == NULL) { head = facts; }
	  tail = facts;
	}
      }
    }
  }
}

double FactsRelErr(double a, double b) {
  double max = (a > b) ? a : b;
  return fabs(a-b)/max;
}

double FactsAbsErr(double a, double b) {
  return fabs(a-b);
}

void FactsPrint(const char *format,...) {
  int i=0,j=0;
  char reformat[80];
  va_list ap;
  va_start(ap, format);
  for (i=0; format[i] != 0;) {
    if (format[i] == '%' && format[i+1] == '?') {
      const char *fs=va_arg(ap, const char *);
      strncpy(reformat+j,fs,80-j);
      j += strlen(fs);
      i += 2;
    } else {
      reformat[j]=format[i];
      ++j; ++i;
    }
    if (j >= sizeof(reformat)) break;
  }
  assert(j < sizeof(reformat)-1);
  reformat[++j]=0;
  vfprintf(stderr,reformat,ap);
  va_end(ap);
}

void FactsCheck() {
  if (head == NULL) { FactsRegister(); }
  for (Facts *facts = head; facts != NULL; facts=facts->next) {
    if (facts->status == 0) {
      fprintf(stderr,"facts %s started\n",facts->name);
      facts->function(facts);
      fprintf(stderr,"facts %s ended%s\n",facts->name,(facts->status == -1 ? " (badly)" : ""));
    } else if (facts->status == -2) {
      fprintf(stderr,"facts %s skipped.\n",facts->name);
    }
  }

  fprintf(stderr,"facts summary.\n");    
  for (Facts *facts = head; facts != NULL; facts=facts->next) {
    if (facts->status == 1) {
	fprintf(stderr,"facts %s passed\n",facts->name);
    }
  }
  for (Facts *facts = head; facts != NULL; facts=facts->next) {
    if (facts->status == -1) {
	fprintf(stderr,"facts %s failed\n",facts->name);
    }
  }
  for (Facts *facts = head; facts != NULL; facts=facts->next) {
    if (facts->status == -2) {
	fprintf(stderr,"facts %s skipped\n",facts->name);
    }
  }
  for (Facts *facts = head; facts != NULL; facts=facts->next) {
    if (facts->status < -2 || 1 < facts->status) {
      fprintf(stderr,"facts %s status %d\n",facts->name,facts->status);
    }
  }
}
