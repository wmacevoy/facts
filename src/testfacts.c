#include "facts.h"

FACTS(Alpha) {
  FACT_PRINT('c',<=,'b',"%c");
}

FACTS(Integer) {
  int a=1,b=2;
  FACT(a,==,b);
}

FACTS(Double) {
  double a=1.3, b=2.4;
  FACT(a,==,b);
}

FACTS_FINISHED

FACTS_MAIN

