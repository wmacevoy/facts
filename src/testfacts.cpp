#include "facts.h"

FACTS(Alpha) {
  FACT('c',<=,'b');
}

FACTS(Integer) {
  int a=1,b=2;
  FACT(a,==,b);
}

FACTS(Double) {
  double a=1.3, b=2.4;
  FACT(a,==,b);
}

FACTS(Err) {
  std::cerr << "message to cerr." << std::endl;
}

FACTS(LargeErr) {
  for (int i=0; i<100; ++i) {
    std::cerr << "message #" << i << " to cerr." << std::endl;
  }
}

FACTS(Out) {
  std::cout << "message to cout." << std::endl;
}

FACTS_FAST
