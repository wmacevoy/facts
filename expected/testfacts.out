src/testfacts.c 3: Alpha facts check started
[1;31msrc/testfacts.c/Alpha 4: 'c' {=c} <= 'b' {=b} is fiction[0m
[1;32mDebug facts_Alpha_function on line 3 of file src/testfacts.c with a breakpoint on line 4. For example in gdb:[0m
break facts_Alpha_function
break "src/testfacts.c":4
run --facts_include Alpha
continue
print 'c'
print 'b'
print ('c') <= ('b')

src/testfacts.c 3: Alpha facts check ended [1;31mbadly[0m
src/testfacts.c 7: Integer facts check started
[1;31msrc/testfacts.c/Integer 9: a {=1} == b {=2} is fiction[0m
[1;32mDebug facts_Integer_function on line 7 of file src/testfacts.c with a breakpoint on line 9. For example in gdb:[0m
break facts_Integer_function
break "src/testfacts.c":9
run --facts_include Integer
continue
print a
print b
print (a) == (b)

src/testfacts.c 7: Integer facts check ended [1;31mbadly[0m
src/testfacts.c 12: Double facts check started
[1;31msrc/testfacts.c/Double 14: a {=1.3} == b {=2.4} is fiction[0m
[1;32mDebug facts_Double_function on line 12 of file src/testfacts.c with a breakpoint on line 14. For example in gdb:[0m
break facts_Double_function
break "src/testfacts.c":14
run --facts_include Double
continue
print a
print b
print (a) == (b)

src/testfacts.c 12: Double facts check ended [1;31mbadly[0m
src/testfacts.c 17: Err facts check started
src/testfacts.c 17: Err facts check ended
src/testfacts.c 21: LargeErr facts check started
src/testfacts.c 21: LargeErr facts check ended
src/testfacts.c 27: Out facts check started
message to stderr.
src/testfacts.c 27: Out facts check ended
facts summary.
facts check Err [1;32mpassed[0m
facts check LargeErr [1;32mpassed[0m
facts check Out [1;32mpassed[0m
facts check Alpha [1;31mfailed[0m
facts check Integer [1;31mfailed[0m
facts check Double [1;31mfailed[0m
0 (0.0%) truths and 3 (100.0%) fictions checked.
